### PR TITLE
[jit] In RPC Server, handle TorchScript continuations asynchronously

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34109 [jit] In RPC Server, handle TorchScript continuations asynchronously**
* #35331 [rref] Handle exceptions returned via remote() calls

This change adds glue to GraphExecutor to give the RPC server
access to the future-based Interpreter::runAsync() api.

Previously, if a server encounted a TorchScript continuation-based block
with fork/wait, it would simply block in the server thread until the handler
completed, since it uses the synchronous Interpreter::run() api.

With the ivalue::Future returned by the Interpreter, we can run the
TorchScript code asynchronously from c++ simply by connecting its
callback to the server callback.

We add test cases to cover the new logic, both rpc_async and remote.

Differential Revision: [D20194321](https://our.internmc.facebook.com/intern/diff/D20194321/)

### CODE REVIEW

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 1]
```diff
@@ -240,6 +241,17 @@ def heavy_rpc_torchscript(tensor):
         tensor /= i + 1
     return 0
 
+
+@torch.jit.script
+def script_sleep_and_add(x):
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: why there is a "sleep" in the name?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 2]
```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
...
-          "Return value of a builtin operator or a "
-          "TorchScript function should be a single IValue, got a vector of "
-          "size ",
-          stack.size());
-
-      return wrap(std::move(ScriptResp(std::move(stack.front()))).toMessage());
+      auto future = PythonRpcHandler::getInstance()
+                        .jitCompilationUnit()
+                        ->get_function(scriptCall.qualifiedName())
+                        .runAsync(stack);
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: Which thread will pick up this async run? Is it some thread from JIT thread pool?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 3]
```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
...
+      if (future->completed()) {
+        return wrap(std::move(ScriptResp(future->value())).toMessage());
+      }
+      auto responseFuture = std::make_shared<FutureMessage>();
+      future->addCallback([responseFuture, messageId, future]() {
+        try {
+          Message m = ScriptResp({future->value()}).toMessage();
+          m.setId(messageId);
+          responseFuture->markCompleted(std::move(m));
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: can we use `wrap` for this as well?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 4]
```diff
@@ -50,23 +50,37 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
...
-          "size ",
-          stack.size());
-
-      return wrap(std::move(ScriptResp(std::move(stack.front()))).toMessage());
+      auto future = PythonRpcHandler::getInstance()
+                        .jitCompilationUnit()
+                        ->get_function(scriptCall.qualifiedName())
+                        .runAsync(stack);
+      if (future->completed()) {
+        return wrap(std::move(ScriptResp(future->value())).toMessage());
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: What is the benefit of separating this case out? Can we consolidate this into `addCallback`? Is the purpose to avoid an additional func invocation?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 5]
```diff
@@ -129,9 +121,48 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
         // addForkOfOwner as it is not a fork. To allow callee to distinguish
         // when this request is sent to self, the caller will set forkId using
         // rrefId (OwnerRRef does not have a forkId anyway).
-        ctx.addForkOfOwner(rrefId, forkId);
+        RRefContext::getInstance().addForkOfOwner(rrefId, forkId);
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: What is the reason for this change?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 6]
```diff
@@ -129,9 +121,48 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
...
+                        ->get_function(scriptRemoteCall.qualifiedName())
+                        .runAsync(stack);
+
+      if (future->completed()) {
+        return std::make_shared<FutureMessage>(onCompletion(future->value()));
+      }
+      auto responseFuture = std::make_shared<FutureMessage>();
+      future->addCallback([responseFuture, future, onCompletion]() {
+        try {
+          responseFuture->markCompleted(onCompletion(future->value()));
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: ditto, can we consolidate with future completed and incomplete cases?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 7]
```diff
@@ -129,9 +121,48 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
...
+      // scriptRemoteCall is only alive within this block, use reference to
+      // avoid copy
+      auto& stack = scriptRemoteCall.stackRef();
+      if (scriptRemoteCall.hasOp()) {
+        scriptRemoteCall.op()->getOperation()(stack);
+        TORCH_INTERNAL_ASSERT(
+            stack.size() == 1,
+            "Return value of a builtin operator or a "
+            "TorchScript function should be a single IValue, got a vector of "
+            "size ",
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: this is prior to this PR, but can we merge these three lines into two?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 2 above)
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: Do we need to worry about the life time of the stack? Will it be possible that the stack is used after the deletion of `scriptCall`?

File: torch/csrc/jit/runtime/graph_executor.cpp [hunk 8]
```diff
@@ -483,6 +483,37 @@ void GraphExecutorImplBase::run(Stack& stack) {
...
+  struct Frame {
+    explicit Frame(ExecutionPlan eplan)
+      : plan(std::move(eplan)),
+        state(plan.code) {
+    }
+    ExecutionPlan plan;
+    InterpreterState state;
+  };
+  auto frame = std::make_shared<Frame>(
+    getPlanFor(stack, GraphExecutor::getDefaultNumBailOuts()));
```
Comment (mrshenli) [inline torch/csrc/jit/runtime/graph_executor.cpp]: Is this how we keep the stack alive?

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 9]
```diff
@@ -1631,3 +1643,19 @@ def test_function_not_on_callee(self):
                 AttributeError, "RPC pickler does not serialize"
             ):
                 rpc.rpc_sync(callee_worker, foo_add, args=())
+
+    @dist_init
+    def test_async_script_udf(self):
+        for i in range(1, 3):
+            future = rpc.rpc_async(
+                "worker{}".format((self.rank + 1) % self.world_size),
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: I guess there is no easy way for us to test if this is indeed launched asynchronously on the server side?

File: torch/testing/_internal/distributed/rpc/rpc_test.py (hunk 1 above)
Comment (jjlilley) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: Sorry, changed the name. This was a vestige of my first attempt, where I tried a sleep, until I realize that we can't time.sleep in TorchScript.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 2 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: It will start the run inline. In most cases, it will simply complete inline as-is.

If there's a torch.jit._wait (see interpreter.cpp:1140), the stack is moved and when the continuation is unblocked, it will run in an at::launch thread.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 3 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Not really:
  - wrap() instantiates a fresh responseFuture with an already-completed value.
  - We already have an existing responseFuture and need to complete it.

We could technically make wrap fill in an existing responseFuture, though it's more efficient not to do so.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 4 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: I expect that probably the majority of the time (i.e. it's just computation, no fork/wait), future will already be completed.

It's more efficient to just instantiate a completed future than to go through the motions of making  an unfilled future, setting up the callback, completing the future, and making the followup callbacks.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 5 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: I just reverted this aspect of the change. It's a vestige of an intermediate version, where I was placing this in a callback and wanted to minimize captures.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 6 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Similar response - let me know if you really don't like it, but consolidating will make the common case less efficient.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 7 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: done

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 2 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: I believe we're fine, please in particular see 
torch/csrc/jit/interpreter.cpp:1143, where we end up effectively move-copying the stack over into the continuation.

File: torch/csrc/jit/runtime/graph_executor.cpp (hunk 8 above)
Comment (jjlilley) [reply torch/csrc/jit/runtime/graph_executor.cpp]: No, this part does not keep the Stack alive, but rather the Code. It's not 100% clear that this was necessary, actually (and actually it would probably be ok not to heap allocate it, but rather stack alloc it and std::move to a heap allocated struct, since the internals are all pimpl-to-shared-ptr, but this felt aggressive)

File: torch/testing/_internal/distributed/rpc/rpc_test.py (hunk 9 above)
Comment (jjlilley) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: Well, I added printfs to convince myself that this was happening, and that we were hitting all the codepaths. But...

If you have ideas for what to throw into python here (particularly that will also work in TorchScript), that would be good, but otherwise, it feels tricky.

**Review by jjlilley — COMMENTED** [review]
Thanks for looking and for the feedback! Uploading a tweaked version

File: torch/testing/_internal/distributed/rpc/rpc_test.py [hunk 10]
```diff
@@ -240,6 +241,17 @@ def heavy_rpc_torchscript(tensor):
         tensor /= i + 1
     return 0
 
+
+@torch.jit.script
```
Comment (xush6528) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: We should put these tests in to "torch/testing/_internal/distributed/rpc/jit/rpc_test.py".

Comment (jjlilley) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: Sure, I'll move it.

I saw both files, but I also saw other torch.jit.script code in the rpc/rpc_test.py file, so I figured that the distinction was fuzzy.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 2 above)
Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: In that case does it mean RPC agent shutdown won't be able to track those threads? If that thread launches additional RPC or use RRef after shutdown, will the errors be propagated correctly?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 4 above)
Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: ah, got it, thanks!

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 2 above)
Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: This probably should be OK, as there is a pending future for it.

**Comment by mrshenli** [comment]
Looks great to me! Thanks @jjlilley! I am just a little timid on using a non-RPC thread.

* What's the implication on RPC debug info? Is there a way for RPC to know how many of those `at::launch` threads are used?
* Shall we add a failure test to make sure that errors are propagated correctly when the user function is executed using `at::launch` thread?

**Comment by jjlilley** [comment]
Thanks for looking.
- I'll add an "async with exceptions" test case and upload in a bit, so that we have test coverage for that codepath.
 
- Re: Rpc Debug Info - we could add a num currently active rpc calls to process_group_agent; let me write that up. 
   I think the num threads stat is going to look less important over time. There's talk with Jerry and folks about consolidating the various thread pools anyway - ideally over time, we can get rid of intermediate non-compute blocking, consolidate the thread pools to avoid thrashing so much between different pools of threads, etc.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 2 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Right, I think process_group_agent's shutdown is implicitly oriented towards "io threads" vs "processing threads"

**Comment by jjlilley** [comment]
@mrshenli , as an update- it was good that you pushed me to write tests for exception support.

This has been a bit hairier than expected (and exposed a few issues in the rest of the plumbing, some related to the disabled test_torchscript_function_exception test, though some of it can be deferred).

I'm getting there, though, hoping to update soon.

**Comment by jjlilley** [comment]
I think this change is ready for review again.
It will merge conflict with my 35331 change (exception propagation through rref), though de-conflicting it will be reasonable, which ever one makes more sense to land first.

**Comment by jjlilley** [comment]
Are there others who I should add to this code review? Thanks!

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 11]
```diff
@@ -178,39 +198,62 @@ void RequestCallbackImpl::processRpc(
 
       auto ownerRRef = ctx.getOrCreateOwnerRRef(rrefId, returnType);
 
-      // TODO: make this asynchronous
-      // scriptRemoteCall is only alive within this block, use reference to
+      auto scriptRemoteCallDone =
+          [ownerRRef, rrefId, forkId, messageId, responseFuture](IValue res) {
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: looks like we no longer need the ownerRRef ptr any more, can we move here?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp [hunk 12]
```diff
@@ -178,39 +198,62 @@ void RequestCallbackImpl::processRpc(
...
+                           .runAsync(stack);
+      if (jitFuture->completed()) {
+        scriptRemoteCallDone(jitFuture->value());
+      } else {
+        jitFuture->addCallback(
+            [scriptRemoteCallDone, jitFuture, responseFuture]() {
+              try {
+                scriptRemoteCallDone(jitFuture->value());
+              } catch (const std::exception& e) {
+                // TODO: the better approach here would be to allow ownerRRef to
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: Since this is added in #35331, do we need to update this accordingly? I am OK with either doing it here or in a followup PR

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks!

**Comment by mrshenli** [comment]
BTW, lint failures should have been fixed on master, could you please rebase?

**Comment by jjlilley** [comment]
Thanks for the review! I'm just finishing running tests locally with a rebase (vs the exception change), and will upload shortly

File: torch/csrc/distributed/rpc/request_callback_impl.cpp (hunk 12 above)
Comment (jjlilley) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: I'm removing this comment, thanks.
